### PR TITLE
Feature/renderer element state

### PR DESCRIPTION
**Pull request by cmeissl**
This PR adds a `OutputRenderReport` that holds the state for each `RenderElement`. The report can be used the query the `RenderElementState` for each element provided to the `render_output` function. The state is usefull to test if an element was skipped or directly scanned out.

The `send_frame` functions in desktop now allow to gate sending the frame callbacks by a provided function.
`desktop` now also offers a function to query the primary output for a window surface.

Anvil has been updated to use the report in combination with the send frame gate function to implement a simple
frame callback throttling for surfaces that have been skipped.
Additionally this also filters the frame callbacks to that they are not driven by the surface primary output.

same goal as #789 but with `wp_presentation_feedback` in mind
~~based on #793~~ merged and rebased

Todo:
- [ ] enhance docs
- [ ] review comments
- [ ] cleanup the mess in `PrimaryScanoutOutput::update_from_output_render_report`
- [ ] better `default_primary_scanout_output_compare` implementation based on refresh rate and visible portion

### CODE REVIEW

File: anvil/src/state.rs [hunk 1]
```diff
@@ -388,15 +390,85 @@ impl<BackendData: Backend + 'static> AnvilState<BackendData> {
...
         for layer_surface in map.layers() {
-            layer_surface.send_frame(self.start_time.elapsed().as_millis() as u32)
+            layer_surface.send_frame(time, |surface, states| {
+                send_throttled_frame(surface, states, report, now, throttle)
+            });
+        }
+    }
+}
+
+fn send_throttled_frame(
```
Comment (Drakulix) [inline anvil/src/state.rs:432]: I find this name a little confusing, given that the frame is only throttled, when the report shows it as skipped.

File: anvil/src/state.rs [hunk 2]
```diff
@@ -388,15 +390,85 @@ impl<BackendData: Backend + 'static> AnvilState<BackendData> {
...
+    states
+        .data_map
+        .insert_if_missing_threadsafe(SurfaceFrameThrottlingState::default);
+    let send_throttled_frame = states
+        .data_map
+        .get::<SurfaceFrameThrottlingState>()
+        .unwrap()
+        .update(tp, throttle);
+
+    !throttle_frame || send_throttled_frame
```
Comment (Drakulix) [inline anvil/src/state.rs:453]: Again the naming, especially with the inverse logic, makes this expression hard to parse.
Something like:
`!was_skipped || frame_overdue` or something along those lines would be easier to understand.

File: anvil/src/state.rs [hunk 3]
```diff
@@ -388,15 +390,85 @@ impl<BackendData: Backend + 'static> AnvilState<BackendData> {
...
+    surface: &WlSurface,
+    states: &smithay::wayland::compositor::SurfaceData,
+    report: &OutputRenderReport,
+    tp: Instant,
+    throttle: Option<Duration>,
+) -> bool {
+    let throttle_frame = report
+        .element_render_state(surface)
+        .map(|state| state == RenderElementState::Skipped)
+        .unwrap_or(true);
```
Comment (Drakulix) [inline anvil/src/state.rs:442]: Maybe event add a function `was_rendered` to the report, that returns false for both `Skipped` or not present elements.

File: src/desktop/space/mod.rs [hunk 4]
```diff
@@ -666,3 +670,54 @@ where
 
     damage_tracked_renderer.render_output(renderer, age, &*render_elements, clear_color, log)
 }
+
+/// Find the primary output for a [`WlSurface`] of a window out of a set of outputs
```
Comment (Drakulix) [inline src/desktop/space/mod.rs:674]: The docs should mention (and maybe the name should be changed to reflect this), that this essentially tries to find the preferred frame rate for a given surface.

For shell-interactions "primary"-output might instead be the output of most overlap, independent of the refresh rate. Using this function for those cases would result in a bad user experience preferring high-refresh outputs over anything else.

That said the function is perfectly fine (and quite clever) for this use-case, it is just not apparent from the naming and documentation, what criteria are used to define "primary".

File: src/desktop/space/mod.rs [hunk 5]
```diff
@@ -668,3 +670,54 @@ where
...
+///
+/// Returns `None` if no primary output could be determined
```
Comment (Drakulix) [inline src/desktop/space/mod.rs:676]: Maybe expand on this. In which cases can this happen? If the output is not "visible" on any output?

What visible means in this context is also questionable. Again using the output state here is quite clever, because it works independently from using either the `Space`-abstraction or driving `output_enter` and `output_leave` manually. But for downstream this is not apparent.

**Review by Drakulix — COMMENTED** [review]
Overall looks very solid, just some remarks on readability/docs of the source code.

File: src/desktop/space/mod.rs (hunk 4 above)
Comment (cmeissl) [reply src/desktop/space/mod.rs:674]: yeah, valid points. I will try to find a different name and document what this functiont does. Maybe it would be a good idea to also provide a function that returns the output of most overlap while at it?

Comment (Drakulix) [reply src/desktop/space/mod.rs:674]: > Maybe it would be a good idea to also provide a function that returns the output of most overlap while at it?

Likely, I have implemented such functionality in both anvil and cosmic-comp at some point, so there is likely a useful helper function

File: anvil/src/state.rs (hunk 1 above)
Comment (cmeissl) [reply anvil/src/state.rs]: renamed to `should_send_frame_callback`

File: anvil/src/state.rs (hunk 2 above)
Comment (cmeissl) [reply anvil/src/state.rs]: now uses `presented || frame_overdue`

File: anvil/src/state.rs (hunk 3 above)
Comment (cmeissl) [reply anvil/src/state.rs]: done, `element_was_presented` function added which returns `false` if not found or skipped

File: src/desktop/space/mod.rs (hunk 4 above)
Comment (cmeissl) [reply src/desktop/space/mod.rs]: renamed to `window_surface_preferred_output` and mentioned the refresh rate in the docs.
also added `window_surface_output_most_overlap` for shell-interactions

File: src/desktop/space/mod.rs (hunk 5 above)
Comment (cmeissl) [reply src/desktop/space/mod.rs]: Updated the docs to include the reason why `None` would be returned and added the requirement of `output_enter`/`output_leave`

**Review by Drakulix — APPROVED** [review]
LGTM

File: anvil/src/state.rs [hunk 6]
```diff
@@ -388,15 +390,83 @@ impl<BackendData: Backend + 'static> AnvilState<BackendData> {
...
+            // as all further checks for the primary output
+            // will fail as well
+            if !outputs.contains(output) {
+                continue;
             }
-        });
+
+            window.send_frame(time, |surface, states| {
+                // We only want to send frame callbacks on the preferred output
+                if !window_surface_preferred_output(surface, &*outputs)
```
Comment (bwidawsk) [inline anvil/src/state.rs:411]: Shouldn't the send_frame only occur after all outputs the surface is shown on have had their part of the surface presented?

File: src/backend/renderer/element/mod.rs [hunk 7]
```diff
@@ -89,6 +96,50 @@ pub enum UnderlyingStorage<'a, R: Renderer> {
...
 }
 
+/// Defines the possible states of an element
+/// after rendering
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum RenderElementState {
+    /// The element was rendered
+    Rendered,
+    /// The element was directly scanned out
+    ScannedOut,
```
Comment (bwidawsk) [inline src/backend/renderer/element/mod.rs:106]: Ideally this would perhaps recorder the presented time. Perhaps `ScannedOut(Duration)`

File: src/backend/renderer/element/mod.rs [hunk 8]
```diff
@@ -89,6 +96,50 @@ pub enum UnderlyingStorage<'a, R: Renderer> {
...
+    /// The element was skipped
+    Skipped,
+}
+
+/// Report for an output render operation
+#[derive(Debug, Clone)]
+pub struct OutputRenderReport {
+    /// Holds the damage for presenting the bound buffer
+    pub damage: Option<Vec<Rectangle<i32, Physical>>>,
```
Comment (bwidawsk) [inline src/backend/renderer/element/mod.rs:115]: I'd drop fallability and just make the empty vec be equivalent to None.

File: src/backend/renderer/element/mod.rs [hunk 9]
```diff
@@ -89,6 +96,50 @@ pub enum UnderlyingStorage<'a, R: Renderer> {
...
+    ///
+    /// Return `None` if the element was not part of the output render operation
+    pub fn element_render_state(&self, id: impl Into<Id>) -> Option<RenderElementState> {
+        self.element_render_states.get(&id.into()).copied()
+    }
+
+    /// Returns whether the element with the specified id was presented
+    ///
+    /// Returns `false` if the element with the id was not found or skipped
+    pub fn element_was_presented(&self, id: impl Into<Id>) -> bool {
```
Comment (bwidawsk) [inline src/backend/renderer/element/mod.rs:136]: This name is confusing IMO. Presented should mean exactly that the state is ScannedOut. I think just `element_was_skipped` should accomplish the same thing afaict.

File: anvil/src/state.rs (hunk 6 above)
Comment (cmeissl) [reply anvil/src/state.rs]: Why? That would limit the frame callbacks to the lowest refresh rate of all outputs it is shown.
The opposite is what you want most of the time, use the output with the highest refresh rate to drive the frame callbacks.

File: src/backend/renderer/element/mod.rs (hunk 7 above)
Comment (cmeissl) [reply src/backend/renderer/element/mod.rs:105]: Okay, totally my fault, this is named pretty badly. That only means it has been assigned to a plane for scan out.

File: src/backend/renderer/element/mod.rs (hunk 8 above)
Comment (cmeissl) [reply src/backend/renderer/element/mod.rs:141]: Right, we could also check for an empty vec

File: src/backend/renderer/element/mod.rs (hunk 9 above)
Comment (cmeissl) [reply src/backend/renderer/element/mod.rs:162]: Same as above, the state represents the rendering and not the scan out, again probably poorly named by me. The check should explicitly return true if the element was not skipped irregardless if the element has been sampled or assigned to a plane.

File: anvil/src/state.rs (hunk 6 above)
Comment (bwidawsk) [reply anvil/src/state.rs]: Looking at this from a power efficiency standpoint, there's no point in starting the client's next draw until all parts of the last one have been displayed.

Suppose you have 2 outputs side by side where the window is spanned across both, one is 120Hz the other is 30Hz. Why would you kick off the next draw faster than 30Hz?

Comment (cmeissl) [reply anvil/src/state.rs]: Because 30 fps will look pretty bad on a 120Hz monitor? Or are least not so nice how it could look like, I mean what is the point in running a monitor with that refresh rate when you limit clients then? The output with the lower refresh rate would sample the buffers at it's refresh rate.
But it may be a valid point for your use case and this is why I struggle what should be put in smithay and what in anvil.

Comment (bwidawsk) [reply anvil/src/state.rs]: Summing up IRC conversation. This is making a choice between power (and overall time where the windows remain in sync), and performance. For the sake of making progress, let's put a pin in it, since I think coverage is a good general heuristic. Perhaps in the future an argument can be added for providing a strategy:
coverage
slowest
fastest

**Comment by cmeissl** [comment]
Just noticed that this architecture in the desktop module will force running the primary scan-out output selection twice when integrating wp_presentation. Maybe a two step approach would be a better fit where the first step runs the selection and the second uses the selected output. So with wp_presentation we would do select output, send frames and send feedback per window? Every function would need to walk the whole surface tree. Or to use a single function instead of send_frames that does send both, frame callbacks and feedback? @Drakulix @bwidawsk opinions?

**Comment by cmeissl** [comment]
> > Or to use a single function instead of send_frames that does send both, frame callbacks and feedback?
> 
> I think I am in favour of this. Generally I would expect every compositor to want to support wp_presentation anyway and not fake any values. (Similar to wp_viewporter.) So coupling the code with this protocol a bit more, should not be problematic, as long as downstream could still decide to not initialize the provided WpPresentation global handler.

Yeah, sounds good. Both things should probably be sent on vblank anyway. If downstream does not initialize wp_presentation there should be no feedback callbacks to send, so that should be fine.

**Review by Drakulix — APPROVED** [review]

File: src/desktop/wayland/layer.rs [hunk 10]
```diff
@@ -562,14 +563,40 @@ impl LayerSurface {
...
+        for (popup, _) in PopupManager::popups_for_surface(surface) {
+            let surface = popup.wl_surface();
+            send_frames_surface_tree(surface, output, time, throttle, primary_scan_out_output);
+        }
+    }
 
-        send_frames_surface_tree(wl_surface, time);
-        for (popup, _) in PopupManager::popups_for_surface(wl_surface) {
-            send_frames_surface_tree(popup.wl_surface(), time);
+    /// Run a closure on all surfaces in this layer
```
Comment (Drakulix) [inline src/desktop/wayland/layer.rs:589]:
```suggestion
    /// Run a closure on all surfaces in this layer (including it's popups if [`PopupManager`] is used)
```

File: src/desktop/wayland/window.rs [hunk 11]
```diff
@@ -218,14 +220,40 @@ impl Window {
...
+        let surface = self.0.toplevel.wl_surface();
+
+        send_frames_surface_tree(surface, output, time, throttle, primary_scan_out_output);
+        for (popup, _) in PopupManager::popups_for_surface(surface) {
+            let surface = popup.wl_surface();
+            send_frames_surface_tree(surface, output, time, throttle, primary_scan_out_output);
+        }
+    }
+
+    /// Run a closure on all surfaces in this window
```
Comment (Drakulix) [inline src/desktop/wayland/window.rs:246]:
```suggestion
    /// Run a closure on all surfaces in this window (including it's popups, if [`PopupManager`] is used)
```

**Review by Drakulix — COMMENTED** [review]
Little nitpick for the docs.

File: anvil/src/udev.rs [hunk 12]
```diff
@@ -935,18 +937,18 @@ fn render_surface<'a>(
         age.into(),
         logger,
     )
-    .map(|x| x.is_some());
+    .map(|(damage, report)| (damage.is_some(), report));
```
[the remark this answers is not in the document]
Comment (cmeissl) [reply anvil/src/udev.rs]: yes, thx

File: src/desktop/wayland/layer.rs (hunk 10 above)
Comment (cmeissl) [reply src/desktop/wayland/layer.rs]: :+1:

File: src/desktop/wayland/window.rs (hunk 11 above)
Comment (cmeissl) [reply src/desktop/wayland/window.rs]: :+1:

**Review by Drakulix — APPROVED** [review]
